### DISCUSSION TITLE
`cargo test` always fails to run.

### DISCUSSION THREAD

**Discussion by HaoYang670** [jorgecarleitao/arrow2#1000] [Q&A] [answered]
Hi all,

When I run `cargo test`, I always get something like
```
error[E0433]: failed to resolve: could not find `parquet` in `io`
 --> examples/parquet_write.rs:9:9
  |
9 |     io::parquet::write::{
  |         ^^^^^^^ could not find `parquet` in `io`
```
Because some features are disabled. 
How could I run `cargo test` successfully?

BTW, `cargo test --features full` always works well.

**Comment by jorgecarleitao** [comment, marked as answer]
Hey @HaoYang670 . So, the idea is: `cargo test --tests` covers everything that is not feature-flagged. `cargo test --lib` is mostly for tests covering functions that are not public (we have very few).

`cargo test` includes examples. Unfortunately `cargo` does not support declaring which feature flags each example requires, and thus we can't provide a `cargo test` that only tests the examples that do not depend on a feature flag.

My usual dev flow is:

If I am working on a core aspect (no feature flags), run `cargo test --tests`. If I am working on a specific feature, run `cargo test --features ... --tests`. I seldom use `cargo test --features full`, since the CI will handle it (and it seldom provides value - if it compiles and the tests under the feature work, very likely all is good.

Ideas on how to improve the situation are welcome, though :)

**Reply by HaoYang670** [reply]
Thank you for your reply @jorgecarleitao !
BTW, maybe we could update the development guide https://github.com/jorgecarleitao/arrow2/blob/main/DEVELOPMENT.md?plain=1#L7-L13 as:
```
## Testing

The simplest way to test the crate is to run 

cargo test --tests

```
